File: src/main/java/com/archid/plugins/store/PluginRepository.java
package com.archid.plugins.store;

import com.archid.plugins.models.Manifest;
import com.archid.plugins.store.entity.PluginEntity;
import com.archid.plugins.store.entity.PluginEntity_;
import io.objectbox.Box;

import java.util.Objects;

public class PluginRepository {
    private final Box<PluginEntity> boxStore;

    public PluginRepository(Box<PluginEntity> boxStore) {
        this.boxStore = boxStore;
    }

    public Manifest findByName(String name) {
        PluginEntity entity = boxStore.query().equal(PluginEntity_.name, name).build().findFirst();

        if (entity == null)
            return null;

        return new Manifest(entity.id, entity.name, entity.classPath, entity.version, entity.enabled);
    }

    public void store(Manifest manifest) {
        if (manifest.getId() <= 0) { // crete
            var entity = new PluginEntity(0, manifest.getName(), manifest.getClassPath(), manifest.getVersion(), manifest.isEnabled());
            boxStore.put(entity);
        } else {
            PluginEntity entity = boxStore.query().equal(PluginEntity_.id, manifest.getId()).build().findFirst();

            if (entity == null)
                entity = boxStore.query().equal(PluginEntity_.name, manifest.getName()).build().findFirst();

            PluginEntity pluginEntity = Objects.requireNonNull(entity);
            pluginEntity.version = manifest.getVersion();
            pluginEntity.enabled = manifest.isEnabled();
            boxStore.put(entity);
        }
    }
}

File: src/main/kotlin/com/archid/plugins/store/PluginRepository.kt
package com.archid.plugins.store

import com.archid.plugins.models.Manifest
import com.archid.plugins.store.entity.PluginManifest
import com.archid.plugins.store.entity.PluginManifests
import org.jetbrains.exposed.sql.Database
import org.jetbrains.exposed.sql.SchemaUtils
import org.jetbrains.exposed.sql.transactions.transaction
import javax.sql.DataSource

class PluginRepository(dataSource: DataSource) {

    init {
        val db = Database.connect(dataSource)
        transaction(db) {
            // Create tables
            SchemaUtils.create(PluginManifests)
        }
    }

    fun findByName(name: String): Manifest? {
        val entity = PluginManifest.find { PluginManifests.name eq name }.firstOrNull() ?: return null

        return Manifest(entity.id.value, entity.name, entity.classPath, entity.version).apply { isEnabled = entity.isEnabled }
    }

    fun store(manifest: Manifest) {
        if (manifest.id <= 0) { // create
            PluginManifest.new {
                name = manifest.name
                classPath = manifest.classPath
                version = manifest.version
                isEnabled = manifest.isEnabled
            }
        } else {
            val entity = PluginManifest.find { PluginManifests.name eq manifest.name }.first()

            entity.version = manifest.version
            entity.isEnabled = manifest.isEnabled
            manifest.id = entity.id.value
        }
    }

}